os-observer/src/events/npm.ts

```diff
@@ -174,67 +196,75 @@ export interface NpmDownloadsArgs {
 export const npmDownloads: EventSourceFunction<NpmDownloadsArgs> = async (
   args: NpmDownloadsArgs,
 ): Promise<ApiReturnType> => {
-  const { name } = args;
+  const { name, autocrawl } = args;
   logger.info(`NPM Downloads: fetching for ${name}`);
 
   // Add the organization and artifact into the database
   const dbArtifact = await getArtifactOrganization(name);
 
   // Get the latest event source pointer
   logger.debug("Getting latest event source pointer");
-  const dbEventSourcePointer = await prisma.eventSourcePointer.findUnique({
-    where: {
-      artifactId_eventType: {
-        artifactId: dbArtifact.id,
-        eventType: EventType.DOWNLOADS,
-      },
-    },
-  });
-  logger.info("EventSourcePointer: ", dbEventSourcePointer);
-  const { lastDate } = safeCast(
-    _.toPlainObject(
-      dbEventSourcePointer?.pointer,
-    ) as Partial<EventSourcePointer>,
+  const previousPointer = await getEventSourcePointer<NpmEventSourcePointer>(
+    dbArtifact.id,
+    EventType.DOWNLOADS,
   );
+  logger.info("EventSourcePointer: ", previousPointer);
 
-  // Retrieve any missing data
   // Start 1 day after the last date we have
-  const start = dayjs(lastDate ?? DEFAULT_START_DATE).add(1, "day");
+  const start = dayjs(previousPointer.lastDate ?? DEFAULT_START_DATE).add(
+    1,
+    "day",
+  );
   // Today's counts may not yet be complete
   const end = dayjs().subtract(1, "day");
+
+  // Short circuit if we're already up to date
+  if (end.isBefore(start)) {
+    return {
+      _type: "upToDate",
+      cached: true,
+    };
+  }
+
+  // Retrieve any missing data
   const downloads: DayDownloads[] = await getDailyDownloads(name, start, end);
 
   // Check for correctness of data
   assert(!hasDuplicates(downloads), "Duplicate dates found in result");
+  const missingDays = getMissingDays(downloads, start, end);
   assert(
-    !hasMissingDays(downloads, start, end),
-    "Missing dates found in result",
+    missingDays.length === 0,
+    `Missing dates found in result: ${missingDays.map(formatDate).join(", ")}`,
   );
 
   // Transform the data into the format we want to store
+  const dbEntries = downloads.map((d) => ({
+    artifactId: dbArtifact.id,
+    eventType: EventType.DOWNLOADS,
+    eventTime: dayjs(d.day).toDate(),
+    amount: d.downloads,
+  }));
 
   // Populate the database
+  await insertData(
+    dbArtifact.id,
+    EventType.DOWNLOADS,
+    dbEntries,
+    previousPointer,
+    safeCast<Partial<NpmEventSourcePointer>>(makeNpmEventSourcePointer(end)),
+    NPM_DOWNLOADS_COMMAND,
+    safeCast<Partial<NpmDownloadsArgs>>(args),
+    autocrawl,
```

Comment:
This feels like the wrong place to be updating autocrawl.

The way I've been thinking about it is we have 2 very different classes of actions: those that move the event pointers forward, and those that do everything else. I think keeping these types of functionality as separate as possible will provide a cleaner interface for people wanting to contribute new event sources.